src/Kis/Kis.hs

```diff
@@ -2,39 +2,69 @@
 {-# LANGUAGE OverloadedStrings          #-}
 {-# LANGUAGE RankNTypes                 #-}
 {-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TypeSynonymInstances       #-}
+{-# LANGUAGE FlexibleInstances          #-}
 
 module Kis.Kis
     ( withKis
-    , KisAction(..)
+    , KisRequest(..)
     , KisClient
     , KisConfig(..)
     , Kis(..)
     , KisException(..)
+    , KisT
+    , createBed
+    , createPatient
+    , getPatients
+    , getPatient
+    , placePatient
+    , runKis
     )
 where
 
 import Control.Monad.Logger
+import Control.Monad.Operational
 import Control.Monad.RWS hiding (asks)
 import Control.Monad.Reader
 import Database.Persist
-import Data.Text
+import Data.Text hiding (singleton)
 import GHC.Exception
 
 import Kis.Model
 import Kis.Time
 
-data KisAction a where
-    CreateBed :: String -> KisAction BedId
-    CreatePatient :: Patient -> KisAction PatientId
-    GetPatient :: PatientId -> KisAction (Maybe Patient)
-    GetPatients :: KisAction [Entity Patient]
-    PlacePatient :: PatientId -> BedId -> KisAction (Maybe PatientBedId)
+data KisRequest a where
+    CreateBed :: String -> KisRequest BedId
+    CreatePatient :: Patient -> KisRequest PatientId
+    GetPatient :: PatientId -> KisRequest (Maybe Patient)
+    GetPatients :: KisRequest [Entity Patient]
+    PlacePatient :: PatientId -> BedId -> KisRequest (Maybe PatientBedId)
 
-deriving instance Show (KisAction a)
+type KisT m = ProgramT KisRequest m
+
+runKis :: KisT IO a -> IO a
+runKis = undefined
+
+class Monad m => MonadKis m where
+    createBed :: String -> m BedId
+    createPatient :: Patient -> m PatientId
+    getPatients :: m [Entity Patient]
+    getPatient :: PatientId -> m (Maybe Patient)
+    placePatient :: PatientId -> BedId -> m (Maybe PatientBedId)
+
+instance Monad m => MonadKis (KisT m) where
+    createPatient = singleton . CreatePatient
+    createBed = singleton . CreateBed
+    getPatients = singleton GetPatients
+    getPatient = singleton . GetPatient
+    placePatient patId bedId = singleton (PlacePatient patId bedId)
+
+
+deriving instance Show (KisRequest a)
 
```

Comment:
Ja, das sieht jetzt erstmal komplexer aus.
Die Typklasse hatten wir vorher halt einfach nicht, sondern wir haben immer direkt in `ReaderT (Kis m) m` gelebt. Man könnte `MonadKis` auch dafür definieren und dann gäbe es die gleiche Duplizierung. Das was in diesem PR `singleton` ist, korrespondiert zu `req` in HEAD. Man könnte auch jetzt immer in `KisT m` bleiben und `singleton (GetPatient patientId)` schreiben, so wie in HEAD `req (GetPatient patientId)`.

Momentan habe ich allerdings die Clock komplett auskommentiert. Ich frage mich, ob die Clock-Funktionalität mit in den GADT sollte oder extra bleibt, also ob man immer Zugriff auf eine Uhr haben will oder das zwei unabhängige Konzepte sind.